pkg/kv/local/store.go

```diff
@@ -0,0 +1,251 @@
+	log := s.logger.
+		WithField("key", string(k)).
+		WithField("op", "set_if").
+		WithContext(ctx)
+	log.Trace("performing operation")
+	if len(partitionKey) == 0 {
+		log.WithError(kv.ErrMissingPartitionKey).Warn("got empty partition key")
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		log.WithError(kv.ErrMissingKey).Warn("got empty key")
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		log.WithError(kv.ErrMissingValue).Warn("got nil value")
+		return kv.ErrMissingValue
+	}
+	err := s.db.Update(func(txn *badger.Txn) error {
+		item, err := txn.Get(k)
+		if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
+			log.WithError(err).Error("could not get key for predicate")
+			return err
+		}
+		if valuePredicate != nil {
+			if errors.Is(err, badger.ErrKeyNotFound) {
+				log.
+					WithField("predicate", nil).
+					Trace("predicate condition failed")
+				return fmt.Errorf("%w: partition=%s, key=%v", kv.ErrPredicateFailed, partitionKey, key)
+			}
+			val, err := item.ValueCopy(nil)
+			if err != nil {
+				log.WithError(err).Error("could not get byte value for predicate")
+				return err
+			}
+			if !bytes.Equal(val, valuePredicate.([]byte)) {
+				log.WithField("predicate", valuePredicate.([]byte)).
+					WithField("value", val).
+					Trace("predicate condition failed")
+				return fmt.Errorf("%w: partition=%s, key=%v", kv.ErrPredicateFailed, partitionKey, key)
+			}
+		} else if !errors.Is(err, badger.ErrKeyNotFound) {
+			log.WithField("predicate", valuePredicate).
+				Trace("predicate condition failed (key not found)")
+			return fmt.Errorf("%w: partition=%s, key=%v", kv.ErrPredicateFailed, partitionKey, key)
+		}
+
+		return txn.Set(composeKey(partitionKey, key), value)
+	})
+	took := time.Since(start)
+	log = log.WithField("took", took)
+	if err != nil {
+		log.WithError(err).Trace("operation failed")
+		return err
+	}
+	log.Trace("operation complete")
+	return nil
+}
+
+func (s *Store) Delete(ctx context.Context, partitionKey, key []byte) error {
+	k := composeKey(partitionKey, key)
+	start := time.Now()
+	log := s.logger.
+		WithField("key", string(k)).
+		WithField("op", "delete").
+		WithContext(ctx)
+	log.Trace("performing operation")
+	if len(partitionKey) == 0 {
+		log.WithError(kv.ErrMissingPartitionKey).Warn("got empty partition key")
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		log.WithError(kv.ErrMissingKey).Warn("got empty key")
+		return kv.ErrMissingKey
+	}
+	err := s.db.Update(func(txn *badger.Txn) error {
+		return txn.Delete(k)
+	})
+	took := time.Since(start)
+	log = log.WithField("took", took)
+	if err != nil {
+		log.WithError(err).Trace("operation failed")
+		return err
+	}
+	log.Trace("operation complete")
+	return nil
+}
+
+func (s *Store) Scan(ctx context.Context, partitionKey, start []byte) (kv.EntriesIterator, error) {
+	var k []byte
+	prefix := partitionRange(partitionKey)
+	if start != nil {
+		k = composeKey(partitionKey, start)
+	} else {
+		k = prefix
+	}
```

Comment:
```suggestion
	k := composeKey(partitionKey, start)
```
I belive it does the same thing

Reply:
We need the prefix - without the start to contain the scan range. modified k initialization

Reply:
I missed the prefix. Thanks